File: stage-gen/src/paths.ts
import { isAbsolute, resolve, sep } from "node:path";
import { lstat, mkdir, realpath } from "node:fs/promises";

const SAFE_SEGMENT = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;

export function assertSafePathSegment(value: string, label: string): string {
  if (!SAFE_SEGMENT.test(value) || value === "." || value === "..") {
    throw new Error(`${label} must be one safe path segment`);
  }
  return value;
}

export function resolveRelativePathWithinRoot(
  rootPath: string,
  requestedPath: string,
  label: string,
): string {
  if (!requestedPath || requestedPath.includes("\0") || isAbsolute(requestedPath)) {
    throw new Error(`${label} must be relative`);
  }
  if (requestedPath.includes("\\")) {
    throw new Error(`${label} contains an invalid path separator`);
  }
  const segments = requestedPath.split("/");
  if (segments.some((segment) => !segment || segment === "." || segment === "..")) {
    throw new Error(`${label} contains an unsafe path segment`);
  }
  const root = resolve(rootPath);
  const output = resolve(root, ...segments);
  if (!output.startsWith(`${root}${sep}`)) throw new Error(`${label} escapes its root`);
  return output;
}

/** Resolve a future output while rejecting symlinked parent directories. */
export async function resolveWritablePathWithinRoot(
  rootPath: string,
  requestedPath: string,
  label: string,
): Promise<string> {
  // Validate before touching the filesystem.
  resolveRelativePathWithinRoot(rootPath, requestedPath, label);
  await mkdir(rootPath, { recursive: true });
  const root = await realpath(rootPath);
  const segments = requestedPath.split("/");
  let parent = root;
  for (const segment of segments.slice(0, -1)) {
    parent = resolve(parent, segment);
    try {
      const stat = await lstat(parent);
      if (stat.isSymbolicLink()) throw new Error(`${label} has a symlinked parent`);
      if (!stat.isDirectory()) throw new Error(`${label} parent is not a directory`);
    } catch (error) {
      if (isMissing(error)) break;
      throw error;
    }
  }
  return resolveRelativePathWithinRoot(root, requestedPath, label);
}

function isMissing(error: unknown): boolean {
  return (
    error !== null &&
    typeof error === "object" &&
    "code" in error &&
    (error as { code?: unknown }).code === "ENOENT"
  );
}

File: src/stage_gen/reliability/paths.py
"""Path confinement helpers for public filesystem boundaries."""

from __future__ import annotations

import os
import re
import stat as stat_module
from pathlib import Path

_SAFE_SEGMENT = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_WINDOWS_ABSOLUTE = re.compile(r"^(?:[A-Za-z]:[\\/]|\\\\)")


def assert_safe_path_segment(value: str, label: str) -> str:
    if not _SAFE_SEGMENT.fullmatch(value) or value in {".", ".."}:
        raise ValueError(f"{label} must be one safe path segment")
    return value


def resolve_relative_path_within_root(
    root_path: str | os.PathLike[str],
    requested_path: str,
    label: str,
) -> Path:
    if (
        not requested_path
        or "\x00" in requested_path
        or Path(requested_path).is_absolute()
        or _WINDOWS_ABSOLUTE.match(requested_path)
    ):
        raise ValueError(f"{label} must be relative")
    if "\\" in requested_path:
        raise ValueError(f"{label} contains an invalid path separator")
    segments = requested_path.split("/")
    if any(not segment or segment in {".", ".."} for segment in segments):
        raise ValueError(f"{label} contains an unsafe path segment")
    root = Path(root_path).absolute()
    output = root.joinpath(*segments).absolute()
    try:
        output.relative_to(root)
    except ValueError as error:
        raise ValueError(f"{label} escapes its root") from error
    return output


def resolve_writable_path_within_root(
    root_path: str | os.PathLike[str],
    requested_path: str,
    label: str,
) -> Path:
    resolve_relative_path_within_root(root_path, requested_path, label)
    root_unresolved = Path(root_path)
    root_unresolved.mkdir(parents=True, exist_ok=True)
    root = root_unresolved.resolve(strict=True)
    segments = requested_path.split("/")
    parent = root
    for segment in segments[:-1]:
        parent = parent / segment
        try:
            stat = parent.lstat()
        except FileNotFoundError:
            break
        if parent.is_symlink():
            raise ValueError(f"{label} has a symlinked parent")
        if not stat_module.S_ISDIR(stat.st_mode):
            raise ValueError(f"{label} parent is not a directory")
    return resolve_relative_path_within_root(root, requested_path, label)
